File: daily_python/1123_Lowest_Common_Ancestor_of_Deepest_Leaves.py
import os
import sys
from typing import List, Optional, Tuple
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Tree:
    def __init__(self, nodes: List[Optional[int]]):
        if not nodes:
            self.root = None
            return
        n = len(nodes)
        i = 1
        self.root = TreeNode(nodes[0])
        que = deque([self.root])
        while i < n and que:
            curr = que.popleft()
            if i < n and nodes[i] is not None:
                curr.left = TreeNode(nodes[i])
                que.append(curr.left)
            i += 1
            if i < n and nodes[i] is not None:
                curr.right = TreeNode(nodes[i])
                que.append(curr.right)
            i += 1

class Solution:
    def dfs(self, node: Optional[TreeNode], depth: int) -> Tuple[Optional[TreeNode], int]:
        if not node:
            return (None, 0)
        left = self.dfs(node.left, depth + 1)
        right = self.dfs(node.right, depth + 1)
        if left[1] > right[1]:
            return left
        elif left[1] < right[1]:
            return right
        else:
            return (node, max(depth, left[1]))
        
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.dfs(root, 1)[0]
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = [None if node == 'None' else int(node) for node in lines[0].split('=')[1].strip()[1:-2].split(',')]
            dataset.append(nodes)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nodes in dataset:
        root = Tree(nodes).root
        results.append(solution.lcaDeepestLeaves(root))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result.val}')

File: daily/2025/1123_Lowest_Common_Ancestor_of_Deepest_Leaves.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* create_tree(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    int n = nodes.size();
    TreeNode* root = new TreeNode(nodes[0]);
    queue<TreeNode*> que;
    que.push(root);
    int i = 1;
    while(!que.empty() && i < n){
        TreeNode* curr = que.front();
        que.pop();
        if (i < n && nodes[i] != -1){
            curr->left = new TreeNode(nodes[i]);
            que.push(curr->left);
        }
        i++;
        if (i < n && nodes[i] != -1){
            curr->right = new TreeNode(nodes[i]);
            que.push(curr->right);
        }
        i++;
    }
    return root;
}

class Solution {
public:
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        return dfs(root, 1).first;
    }
    pair<TreeNode*, int> dfs(TreeNode* root, int depth){
        if (!root) return {nullptr, 0};
        pair<TreeNode*, int> left = dfs(root->left, depth+1);
        pair<TreeNode*, int> right = dfs(root->right, depth+1);
        if (left.second > right.second) return left;
        else if (right.second > left.second) return right;
        else return {root, max(depth,left.second)};
        return {nullptr, -1};
    }
};

int main(){
    vector<int> nodes = {3,5,1,6,2,0,8,-1,-1,7,4};
    // vector<int> nodes = {1};
    // vector<int> nodes = {0,1,3,-1,2};
    TreeNode* root = create_tree(nodes);
    Solution S;
    cout << S.lcaDeepestLeaves(root)->val << endl;
    return 0;
}